Header inclusion fails on AVX512

Compile the following program
```
#include <xsimd/xsimd.hpp>
```
with command
```
/usr/bin/c++  -DXTENSOR_USE_XSIMD -isystem /where-xsimd-lives/include -g   -Wall -Wconversion -Wsign-compare -Wno-sign-conversion -fdiagnostics-color -march=skylake-avx512 -msse2 -msse3 -mssse3 -msse4.1 -msse4.2 -mavx -mfma -mbmi2 -mavx2 -mavx512f -mavx512vl -mavx512cd -mavx512dq -mavx512bw -mno-sse4a -mno-xop -mno-fma4 -mno-avx512pf -mno-avx512er -mno-avx512ifma -mno-avx512vbmi -std=c++17 -o a.o -c a.cc
```
fails with
```
In file included from /usr/lib/gcc/x86_64-linux-gnu/8/include/immintrin.h:55,
                 from /usr/lib/gcc/x86_64-linux-gnu/8/include/x86intrin.h:48,
                 from /where-xsimd-lives/g/xsimd/config/xsimd_include.hpp:17,
                 from /where-xsimd-lives/g/xsimd/types/xsimd_types_include.hpp:12,
                 from /where-xsimd-lives/g/xsimd/types/xsimd_traits.hpp:15,
                 from /where-xsimd-lives/g/xsimd/xsimd.hpp:14,
                 from a.cc:1:
/where-xsimd-lives/g/xsimd/types/xsimd_avx512_int16.hpp: In instantiation of 'static xsimd::detail::avx512_int16_batch_kernel<T>::batch_type xsimd::detail::avx512_int16_batch_kernel<T>::select(const batch_bool_type&, const batch_type&, const batch_type&) [with T = short int; xsimd::detail::avx512_int16_batch_kernel<T>::batch_type = xsimd::batch<short int, 32>; xsimd::detail::avx512_int16_batch_kernel<T>::batch_bool_type = xsimd::batch_bool<short int, 32>]':
/where-xsimd-lives/g/xsimd/types/xsimd_base.hpp:1623:30:   required from 'xsimd::batch_type_t<X> xsimd::select(const typename xsimd::simd_batch_traits<X>::batch_bool_type&, const xsimd::simd_base<X>&, const xsimd::simd_base<X>&) [with X = xsimd::batch<short int, 32>; xsimd::batch_type_t<X> = xsimd::batch<short int, 32>; typename xsimd::simd_batch_traits<X>::batch_bool_type = xsimd::batch_bool<short int, 32>]'
/where-xsimd-lives/g/xsimd/types/xsimd_avx512_conversion.hpp:45:5:   required from here
/where-xsimd-lives/g/xsimd/types/xsimd_avx512_int16.hpp:313:24: error: can't convert value to a vector
                 return _mm512_mask_blend_epi16(cond, b, a);
                        ^~~~~~~~~~~~~~~~~~~~~~~
/where-xsimd-lives/g/xsimd/types/xsimd_avx512_int16.hpp:313:24: error: can't convert value to a vector
                 return _mm512_mask_blend_epi16(cond, b, a);
                        ^~~~~~~~~~~~~~~~~~~~~~~

```

GCC 7.4 also has the same error 😞 
```
-- The C compiler identification is GNU 7.4.0
-- The CXX compiler identification is GNU 7.4.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done
-- Generating done

```

```
In file included from /usr/lib/gcc/x86_64-linux-gnu/7/include/immintrin.h:55:0,
                 from /usr/lib/gcc/x86_64-linux-gnu/7/include/x86intrin.h:48,
                 from /usr/local/include/xsimd/config/xsimd_include.hpp:19,
                 from /usr/local/include/xsimd/types/xsimd_types_include.hpp:14,
                 from /usr/local/include/xsimd/types/xsimd_traits.hpp:17,
                 from /usr/local/include/xsimd/xsimd.hpp:16,
                 from /home/linux/simd/main.cpp:2:
/usr/local/include/xsimd/types/xsimd_avx512_int16.hpp: In instantiation of 'static xsimd::detail::avx512_int16_batch_kernel<T>::batch_type xsimd::detail::avx512_int16_batch_kernel<T>::select(const batch_bool_type&, const batch_type&, const batch_type&) [with T = short int; xsimd::detail::avx512_int16_batch_kernel<T>::batch_type = xsimd::batch<short int, 32>; xsimd::detail::avx512_int16_batch_kernel<T>::batch_bool_type = xsimd::batch_bool<short int, 32>]':
/usr/local/include/xsimd/types/xsimd_base.hpp:1625:30:   required from 'xsimd::batch_type_t<X> xsimd::select(const typename xsimd::simd_batch_traits<X>::batch_bool_type&, const xsimd::simd_base<X>&, const xsimd::simd_base<X>&) [with X = xsimd::batch<short int, 32>; xsimd::batch_type_t<X> = xsimd::batch<short int, 32>; typename xsimd::simd_batch_traits<X>::batch_bool_type = xsimd::batch_bool<short int, 32>]'
/usr/local/include/xsimd/types/xsimd_avx512_conversion.hpp:47:5:   required from here
/usr/local/include/xsimd/types/xsimd_avx512_int16.hpp:315:24: error: can't convert value to a vector
                 return _mm512_mask_blend_epi16(cond, b, a);
                        ^
/usr/local/include/xsimd/types/xsimd_avx512_int16.hpp:315:24: error: can't convert value to a vector
                 return _mm512_mask_blend_epi16(cond, b, a);
                        ^
/usr/local/include/xsimd/types/xsimd_avx512_int16.hpp: In instantiation of 'static xsimd::detail::avx512_int16_batch_kernel<T>::batch_type xsimd::detail::avx512_int16_batch_kernel<T>::select(const batch_bool_type&, const batch_type&, const batch_type&) [with T = short unsigned int; xsimd::detail::avx512_int16_batch_kernel<T>::batch_type = xsimd::batch<short unsigned int, 32>; xsimd::detail::avx512_int16_batch_kernel<T>::batch_bool_type = xsimd::batch_bool<short unsigned int, 32>]':
/usr/local/include/xsimd/types/xsimd_base.hpp:1625:30:   required from 'xsimd::batch_type_t<X> xsimd::select(const typename xsimd::simd_batch_traits<X>::batch_bool_type&, const xsimd::simd_base<X>&, const xsimd::simd_base<X>&) [with X = xsimd::batch<short unsigned int, 32>; xsimd::batch_type_t<X> = xsimd::batch<short unsigned int, 32>; typename xsimd::simd_batch_traits<X>::batch_bool_type = xsimd::batch_bool<short unsigned int, 32>]'
/usr/local/include/xsimd/types/xsimd_avx512_conversion.hpp:47:5:   required from here
/usr/local/include/xsimd/types/xsimd_avx512_int16.hpp:315:24: error: can't convert value to a vector
                 return _mm512_mask_blend_epi16(cond, b, a);
                        ^
/usr/local/include/xsimd/types/xsimd_avx512_int16.hpp:315:24: error: can't convert value to a vector
                 return _mm512_mask_blend_epi16(cond, b, a);
                        ^
/usr/local/include/xsimd/types/xsimd_avx512_int8.hpp: In instantiation of 'static xsimd::detail::avx512_int8_batch_kernel<T>::batch_type xsimd::detail::avx512_int8_batch_kernel<T>::select(const batch_bool_type&, const batch_type&, const batch_type&) [with T = signed char; xsimd::detail::avx512_int8_batch_kernel<T>::batch_type = xsimd::batch<signed char, 64>; xsimd::detail::avx512_int8_batch_kernel<T>::batch_bool_type = xsimd::batch_bool<signed char, 64>]':
/usr/local/include/xsimd/types/xsimd_base.hpp:1625:30:   required from 'xsimd::batch_type_t<X> xsimd::select(const typename xsimd::simd_batch_traits<X>::batch_bool_type&, const xsimd::simd_base<X>&, const xsimd::simd_base<X>&) [with X = xsimd::batch<signed char, 64>; xsimd::batch_type_t<X> = xsimd::batch<signed char, 64>; typename xsimd::simd_batch_traits<X>::batch_bool_type = xsimd::batch_bool<signed char, 64>]'
/usr/local/include/xsimd/types/xsimd_avx512_conversion.hpp:47:5:   required from here
/usr/local/include/xsimd/types/xsimd_avx512_int8.hpp:317:24: error: can't convert value to a vector
                 return _mm512_mask_blend_epi8(cond, b, a);
                        ^
/usr/local/include/xsimd/types/xsimd_avx512_int8.hpp:317:24: error: can't convert value to a vector
                 return _mm512_mask_blend_epi8(cond, b, a);
                        ^
/usr/local/include/xsimd/types/xsimd_avx512_int8.hpp: In instantiation of 'static xsimd::detail::avx512_int8_batch_kernel<T>::batch_type xsimd::detail::avx512_int8_batch_kernel<T>::select(const batch_bool_type&, const batch_type&, const batch_type&) [with T = unsigned char; xsimd::detail::avx512_int8_batch_kernel<T>::batch_type = xsimd::batch<unsigned char, 64>; xsimd::detail::avx512_int8_batch_kernel<T>::batch_bool_type = xsimd::batch_bool<unsigned char, 64>]':
/usr/local/include/xsimd/types/xsimd_base.hpp:1625:30:   required from 'xsimd::batch_type_t<X> xsimd::select(const typename xsimd::simd_batch_traits<X>::batch_bool_type&, const xsimd::simd_base<X>&, const xsimd::simd_base<X>&) [with X = xsimd::batch<unsigned char, 64>; xsimd::batch_type_t<X> = xsimd::batch<unsigned char, 64>; typename xsimd::simd_batch_traits<X>::batch_bool_type = xsimd::batch_bool<unsigned char, 64>]'
/usr/local/include/xsimd/types/xsimd_avx512_conversion.hpp:47:5:   required from here
/usr/local/include/xsimd/types/xsimd_avx512_int8.hpp:317:24: error: can't convert value to a vector
                 return _mm512_mask_blend_epi8(cond, b, a);
                        ^
/usr/local/include/xsimd/types/xsimd_avx512_int8.hpp:317:24: error: can't convert value to a vector
                 return _mm512_mask_blend_epi8(cond, b, a);

```

```
In file included from /usr/lib/gcc/x86_64-linux-gnu/8/include/immintrin.h:55,
                 from /usr/lib/gcc/x86_64-linux-gnu/8/include/x86intrin.h:48,
                 from /usr/local/include/xsimd/config/xsimd_include.hpp:19,
                 from /usr/local/include/xsimd/types/xsimd_types_include.hpp:14,
                 from /usr/local/include/xsimd/types/xsimd_traits.hpp:17,
                 from /usr/local/include/xsimd/xsimd.hpp:16,
                 from main.cpp:1:
/usr/local/include/xsimd/types/xsimd_avx512_int16.hpp: In instantiation of ‘static xsimd::detail::avx512_int16_batch_kernel<T>::batch_type xsimd::detail::avx512_int16_batch_kernel<T>::select(const batch_bool_type&, const batch_type&, const batch_type&) [with T = short int; xsimd::detail::avx512_int16_batch_kernel<T>::batch_type = xsimd::batch<short int, 32>; xsimd::detail::avx512_int16_batch_kernel<T>::batch_bool_type = xsimd::batch_bool<short int, 32>]’:
/usr/local/include/xsimd/types/xsimd_base.hpp:1625:30:   required from ‘xsimd::batch_type_t<X> xsimd::select(const typename xsimd::simd_batch_traits<X>::batch_bool_type&, const xsimd::simd_base<X>&, const xsimd::simd_base<X>&) [with X = xsimd::batch<short int, 32>; xsimd::batch_type_t<X> = xsimd::batch<short int, 32>; typename xsimd::simd_batch_traits<X>::batch_bool_type = xsimd::batch_bool<short int, 32>]’
/usr/local/include/xsimd/types/xsimd_avx512_conversion.hpp:47:5:   required from here
/usr/local/include/xsimd/types/xsimd_avx512_int16.hpp:319:18: error: can’t convert value to a vector
   __m512i mres = _mm512_mask_blend_epi16(cond, b, a);
                  ^~~~~~~~~~~~~~~~~~~~~~~
/usr/local/include/xsimd/types/xsimd_avx512_int16.hpp:319:18: error: can’t convert value to a vector
   __m512i mres = _mm512_mask_blend_epi16(cond, b, a);
                  ^~~~~~~~~~~~~~~~~~~~~~~
/usr/local/include/xsimd/types/xsimd_avx512_int16.hpp: In instantiation of ‘static xsimd::detail::avx512_int16_batch_kernel<T>::batch_type xsimd::detail::avx512_int16_batch_kernel<T>::select(const batch_bool_type&, const batch_type&, const batch_type&) [with T = short unsigned int; xsimd::detail::avx512_int16_batch_kernel<T>::batch_type = xsimd::batch<short unsigned int, 32>; xsimd::detail::avx512_int16_batch_kernel<T>::batch_bool_type = xsimd::batch_bool<short unsigned int, 32>]’:
/usr/local/include/xsimd/types/xsimd_base.hpp:1625:30:   required from ‘xsimd::batch_type_t<X> xsimd::select(const typename xsimd::simd_batch_traits<X>::batch_bool_type&, const xsimd::simd_base<X>&, const xsimd::simd_base<X>&) [with X = xsimd::batch<short unsigned int, 32>; xsimd::batch_type_t<X> = xsimd::batch<short unsigned int, 32>; typename xsimd::simd_batch_traits<X>::batch_bool_type = xsimd::batch_bool<short unsigned int, 32>]’
/usr/local/include/xsimd/types/xsimd_avx512_conversion.hpp:47:5:   required from here
/usr/local/include/xsimd/types/xsimd_avx512_int16.hpp:319:18: error: can’t convert value to a vector
   __m512i mres = _mm512_mask_blend_epi16(cond, b, a);
                  ^~~~~~~~~~~~~~~~~~~~~~~
/usr/local/include/xsimd/types/xsimd_avx512_int16.hpp:319:18: error: can’t convert value to a vector
   __m512i mres = _mm512_mask_blend_epi16(cond, b, a);
                  ^~~~~~~~~~~~~~~~~~~~~~~
/usr/local/include/xsimd/types/xsimd_avx512_int8.hpp: In instantiation of ‘static xsimd::detail::avx512_int8_batch_kernel<T>::batch_type xsimd::detail::avx512_int8_batch_kernel<T>::select(const batch_bool_type&, const batch_type&, const batch_type&) [with T = signed char; xsimd::detail::avx512_int8_batch_kernel<T>::batch_type = xsimd::batch<signed char, 64>; xsimd::detail::avx512_int8_batch_kernel<T>::batch_bool_type = xsimd::batch_bool<signed char, 64>]’:
/usr/local/include/xsimd/types/xsimd_base.hpp:1625:30:   required from ‘xsimd::batch_type_t<X> xsimd::select(const typename xsimd::simd_batch_traits<X>::batch_bool_type&, const xsimd::simd_base<X>&, const xsimd::simd_base<X>&) [with X = xsimd::batch<signed char, 64>; xsimd::batch_type_t<X> = xsimd::batch<signed char, 64>; typename xsimd::simd_batch_traits<X>::batch_bool_type = xsimd::batch_bool<signed char, 64>]’
/usr/local/include/xsimd/types/xsimd_avx512_conversion.hpp:47:5:   required from here
/usr/local/include/xsimd/types/xsimd_avx512_int8.hpp:320:18: error: can’t convert value to a vector
   __m512i mres = _mm512_mask_blend_epi8(cond, b, a);
                  ^~~~~~~~~~~~~~~~~~~~~~
/usr/local/include/xsimd/types/xsimd_avx512_int8.hpp:320:18: error: can’t convert value to a vector
   __m512i mres = _mm512_mask_blend_epi8(cond, b, a);
                  ^~~~~~~~~~~~~~~~~~~~~~
/usr/local/include/xsimd/types/xsimd_avx512_int8.hpp: In instantiation of ‘static xsimd::detail::avx512_int8_batch_kernel<T>::batch_type xsimd::detail::avx512_int8_batch_kernel<T>::select(const batch_bool_type&, const batch_type&, const batch_type&) [with T = unsigned char; xsimd::detail::avx512_int8_batch_kernel<T>::batch_type = xsimd::batch<unsigned char, 64>; xsimd::detail::avx512_int8_batch_kernel<T>::batch_bool_type = xsimd::batch_bool<unsigned char, 64>]’:
/usr/local/include/xsimd/types/xsimd_base.hpp:1625:30:   required from ‘xsimd::batch_type_t<X> xsimd::select(const typename xsimd::simd_batch_traits<X>::batch_bool_type&, const xsimd::simd_base<X>&, const xsimd::simd_base<X>&) [with X = xsimd::batch<unsigned char, 64>; xsimd::batch_type_t<X> = xsimd::batch<unsigned char, 64>; typename xsimd::simd_batch_traits<X>::batch_bool_type = xsimd::batch_bool<unsigned char, 64>]’
/usr/local/include/xsimd/types/xsimd_avx512_conversion.hpp:47:5:   required from here
/usr/local/include/xsimd/types/xsimd_avx512_int8.hpp:320:18: error: can’t convert value to a vector
   __m512i mres = _mm512_mask_blend_epi8(cond, b, a);
                  ^~~~~~~~~~~~~~~~~~~~~~
/usr/local/include/xsimd/types/xsimd_avx512_int8.hpp:320:18: error: can’t convert value to a vector
   __m512i mres = _mm512_mask_blend_epi8(cond, b, a);

```

`__m512i mres = _mm512_mask_blend_epi16(cond, mb, ma);` works
`__m512i mres = _mm512_mask_blend_epi16(c, b, a);` throws error
```
/usr/local/include/xsimd/types/xsimd_avx512_int16.hpp: In instantiation of ‘static xsimd::detail::avx512_int16_batch_kernel<T>::batch_type xsimd::detail::avx512_int16_batch_kernel<T>::select(const batch_bool_type&, const batch_type&, const batch_type&) [with T = short int; xsimd::detail::avx512_int16_batch_kernel<T>::batch_type = xsimd::batch<short int, 32>; xsimd::detail::avx512_int16_batch_kernel<T>::batch_bool_type = xsimd::batch_bool<short int, 32>]’:
/usr/local/include/xsimd/types/xsimd_base.hpp:1625:30:   required from ‘xsimd::batch_type_t<X> xsimd::select(const typename xsimd::simd_batch_traits<X>::batch_bool_type&, const xsimd::simd_base<X>&, const xsimd::simd_base<X>&) [with X = xsimd::batch<short int, 32>; xsimd::batch_type_t<X> = xsimd::batch<short int, 32>; typename xsimd::simd_batch_traits<X>::batch_bool_type = xsimd::batch_bool<short int, 32>]’
/usr/local/include/xsimd/types/xsimd_avx512_conversion.hpp:47:5:   required from here
/usr/local/include/xsimd/types/xsimd_avx512_int16.hpp:319:18: error: can’t convert value to a vector
   __m512i mres = _mm512_mask_blend_epi16(c, b, a);
                  ^~~~~~~~~~~~~~~~~~~~~~~
/usr/local/include/xsimd/types/xsimd_avx512_int16.hpp:319:18: error: can’t convert value to a vector
   __m512i mres = _mm512_mask_blend_epi16(c, b, a);
                  ^~~~~~~~~~~~~~~~~~~~~~~

```

--- a/include/xsimd/types/xsimd_avx512_int16.hpp
+++ b/include/xsimd/types/xsimd_avx512_int16.hpp
@@ -312,7 +312,11 @@ namespace xsimd
             static batch_type select(const batch_bool_type& cond, const batch_type& a, const batch_type& b)
             {
             #if defined(XSIMD_AVX512BW_AVAILABLE)
-                return _mm512_mask_blend_epi16(cond, b, a);
+                // Some compilers are not happy with passing directly a and b to the intrinsics
+                // See https://github.com/xtensor-stack/xsimd/issues/315
+                __m512i ma = a;
+                __m512i mb = b;
+                return _mm512_mask_blend_epi16(cond, mb, ma);
             #else
                 XSIMD_SPLIT_AVX512(cond);
                 XSIMD_SPLIT_AVX512(a);

--- a/include/xsimd/types/xsimd_avx512_int8.hpp
+++ b/include/xsimd/types/xsimd_avx512_int8.hpp
@@ -314,7 +314,11 @@ namespace xsimd
             static batch_type select(const batch_bool_type& cond, const batch_type& a, const batch_type& b)
             {
             #if defined(XSIMD_AVX512BW_AVAILABLE)
-                return _mm512_mask_blend_epi8(cond, b, a);
+                // Some compilers are not happy with passing directly a and b to the intrinsics
+                // See https://github.com/xtensor-stack/xsimd/issues/315
+                __m512i ma = a;
+                __m512i mb = b;
+                return _mm512_mask_blend_epi8(cond, mb, ma);
             #else
                 XSIMD_SPLIT_AVX512(cond);
                 XSIMD_SPLIT_AVX512(a);

